<a href="https://colab.research.google.com/github/Capechusami/CSEC_Data_Science_Bootcamp-Task_3/blob/main/Task_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part 1: Create Your Own Dataset (Foundation)

In this task, we will create a **custom dataset using pandas**.  
The requirements are:
- 5 columns (features)
- 15 rows
- Custom index

We will use a **student performance dataset** with the following columns:
1. Name
2. Age
3. Department
4. Score
5. Graduated

In [1]:
# Import pandas library
import pandas as pd

## Step 1: Define a Custom Index

We will create 15 unique labels for our rows, from `S1` to `S15`.

In [2]:
# Custom index for 15 rows
index_labels = [
    "S1", "S2", "S3", "S4", "S5",
    "S6", "S7", "S8", "S9", "S10",
    "S11", "S12", "S13", "S14", "S15"
]

## Step 2: Create the Dataset

We will define our 5 columns using Python lists and combine them into a pandas DataFrame.

In [3]:
# Create the DataFrame
df = pd.DataFrame({
    "Name": [
        "Alice", "Bob", "Charlie", "David", "Eva",
        "Frank", "Grace", "Hannah", "Ian", "Jane",
        "Kevin", "Lily", "Mike", "Nina", "Oscar"
    ],
    "Age": [
        20, 21, 22, 23, 20,
        21, 22, 23, 24, 20,
        21, 22, 23, 24, 25
    ],
    "Department": [
        "CS", "Math", "Physics", "CS", "Math",
        "Physics", "CS", "Math", "Physics", "CS",
        "Math", "Physics", "CS", "Math", "Physics"
    ],
    "Score": [
        85, 78, 92, 88, 76,
        95, 89, 84, 91, 87,
        79, 93, 90, 82, 88
    ],
    "Graduated": [
        "No", "No", "Yes", "Yes", "No",
        "Yes", "Yes", "No", "Yes", "No",
        "No", "Yes", "Yes", "No", "Yes"
    ]
}, index=index_labels)

## Step 3: Display the Dataset

Now we will display the dataset to verify that it meets all requirements:
- 5 columns
- 15 rows
- Custom index

In [4]:
# Display the DataFrame
df

,Name,Age,Department,Score,Graduated
S1,Alice,20,CS,85,No
S2,Bob,21,Math,78,No
S3,Charlie,22,Physics,92,Yes
S4,David,23,CS,88,Yes
S5,Eva,20,Math,76,No
S6,Frank,21,Physics,95,Yes
S7,Grace,22,CS,89,Yes
S8,Hannah,23,Math,84,No
S9,Ian,24,Physics,91,Yes
S10,Jane,20,CS,87,No


# Part 2: Titanic Dataset Analysis

In this part, we will perform a **complete data science workflow** on the Titanic dataset (`train.csv`).  
The steps include:

1. Data Exploration
2. Data Cleaning
3. Data Analysis using `groupby`
4. Filtering specific groups
5. Drawing Insights

In [5]:
# Import pandas
import pandas as pd

## Step 1: Load Dataset

We will load the Titanic dataset (`train.csv`).  


In [8]:
# Load the dataset
df = pd.read_csv("train.csv")

# Display the first 5 rows
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Step 2: Data Exploration

We use:
- `.head()` → preview first rows
- `.info()` → check structure, data types, missing values
- `.describe()` → statistics for numerical columns

In [9]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [11]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## Step 3: Data Cleaning

Tasks:
1. Fill missing values:
   - `Age` → median
   - `Embarked` → mode
2. Drop `Cabin` column (too many missing values)
3. Remove duplicate rows

In [12]:
# Fill missing Age with median
df["Age"] = df["Age"].fillna(df["Age"].median())

# Fill missing Embarked with mode
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])

# Drop Cabin column
df = df.drop(columns=["Cabin"])

# Remove duplicates
df = df.drop_duplicates()

# Verify cleaning
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


## Step 4: Data Analysis

We use **`groupby`** to analyze survival rates:

- By gender
- By class
- Average age per class
- By age group

In [13]:
# Survival rate by gender
df.groupby("Sex")["Survived"].mean()

,Survived
Sex,
female,0.742038
male,0.188908


In [14]:
# Survival rate by class
df.groupby("Pclass")["Survived"].mean()

,Survived
Pclass,
1,0.629630
2,0.472826
3,0.242363


In [15]:
# Average age per class
df.groupby("Pclass")["Age"].mean()

,Age
Pclass,
1,36.812130
2,29.765380
3,25.932627


In [16]:
# Create Age Groups: Child, Teen, Adult, Senior
df["AgeGroup"] = pd.cut(
    df["Age"],
    bins=[0, 12, 18, 60, 100],
    labels=["Child", "Teen", "Adult", "Senior"]
)

# Survival rate by age group
df.groupby("AgeGroup")["Survived"].mean()

/tmp/ipykernel_2406/3067060114.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby("AgeGroup")["Survived"].mean()


,Survived
AgeGroup,
Child,0.579710
Teen,0.428571
Adult,0.365753
Senior,0.227273


## Step 5: Filtering Specific Groups

We filter:

1. Female passengers who survived
2. Children who survived
3. 1st class passengers who survived

In [17]:
# Female survivors
df[(df["Sex"] == "female") & (df["Survived"] == 1)]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Adult
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,Adult
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Adult
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,S,Adult
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,C,Teen
...,...,...,...,...,...,...,...,...,...,...,...,...
874,875,1,2,"Abelson, Mrs. Samuel (Hannah Wizosky)",female,28.0,1,0,P/PP 3381,24.0000,C,Adult
875,876,1,3,"Najib, Miss. Adele Kiamie ""Jane""",female,15.0,0,0,2667,7.2250,C,Teen
879,880,1,1,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",female,56.0,0,1,11767,83.1583,C,Adult
880,881,1,2,"Shelley, Mrs. William (Imanita Parrish Hall)",female,25.0,0,1,230433,26.0000,S,Adult


In [18]:
# Children survivors (<12 years old)
df[(df["Age"] < 12) & (df["Survived"] == 1)]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup
10,11,1,3,"Sandstrom, Miss. Marguerite Rut",female,4.00,1,1,PP 9549,16.7000,S,Child
43,44,1,2,"Laroche, Miss. Simonne Marie Anne Andree",female,3.00,1,2,SC/Paris 2123,41.5792,C,Child
58,59,1,2,"West, Miss. Constance Mirium",female,5.00,1,2,C.A. 34651,27.7500,S,Child
78,79,1,2,"Caldwell, Master. Alden Gates",male,0.83,0,2,248738,29.0000,S,Child
165,166,1,3,"Goldsmith, Master. Frank John William ""Frankie""",male,9.00,0,2,363291,20.5250,S,Child
172,173,1,3,"Johnson, Miss. Eleanor Ileen",female,1.00,1,1,347742,11.1333,S,Child
183,184,1,2,"Becker, Master. Richard F",male,1.00,2,1,230136,39.0000,S,Child
184,185,1,3,"Kink-Heilmann, Miss. Luise Gretchen",female,4.00,0,2,315153,22.0250,S,Child
193,194,1,2,"Navratil, Master. Michel M",male,3.00,1,1,230080,26.0000,S,Child
233,234,1,3,"Asplund, Miss. Lillian Gertrud",female,5.00,4,2,347077,31.3875,S,Child


In [19]:
# 1st Class survivors
df[(df["Pclass"] == 1) & (df["Survived"] == 1)]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,AgeGroup
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Adult
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Adult
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,S,Adult
23,24,1,1,"Sloper, Mr. William Thompson",male,28.0,0,0,113788,35.5000,S,Adult
31,32,1,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,28.0,1,0,PC 17569,146.5208,C,Adult
...,...,...,...,...,...,...,...,...,...,...,...,...
862,863,1,1,"Swift, Mrs. Frederick Joel (Margaret Welles Ba...",female,48.0,0,0,17466,25.9292,S,Adult
871,872,1,1,"Beckwith, Mrs. Richard Leonard (Sallie Monypeny)",female,47.0,1,1,11751,52.5542,S,Adult
879,880,1,1,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",female,56.0,0,1,11767,83.1583,C,Adult
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S,Adult


# Step 6: Insights – Detailed Explanation

This step is where we interpret the results of our analysis to answer the key questions about Titanic passengers' survival.



## 1. Who was more likely to survive?

**Females had a much higher survival rate than males.**

**How we know:** We calculated survival rates by gender: `df.groupby("Sex")["Survived"].mean()`

**Example result:**
| Sex    | Survived |
|--------|----------|
| female | 0.74     |
| male   | 0.19     |

**Interpretation:** 74% of females survived compared to only 19% of males. This huge difference reflects the "women and children first" policy on the Titanic. Even without looking at class or age, gender alone is a strong predictor of survival.





## 2. Did class affect survival?

**Yes, 1st class passengers had the highest survival rates.**

**How we know:** Survival by passenger class was calculated: `df.groupby("Pclass")["Survived"].mean()`

**Example result:**
| Pclass | Survived |
|--------|----------|
| 1      | 0.63     |
| 2      | 0.47     |
| 3      | 0.24     |

**Interpretation:** 1st class passengers had 63% survival, 2nd class 47%, and 3rd class only 24%. This suggests access to lifeboats was easier for higher-class passengers. 3rd class passengers were farther from lifeboats and below deck, leading to lower survival.




## 3. Were children prioritized?

**Yes, children had higher survival rates than adults.**

**How we know:** We created age groups: `df["AgeGroup"] = pd.cut(df["Age"], bins=[0,12,18,60,100], labels=["Child","Teen","Adult","Senior"])` then `df.groupby("AgeGroup")["Survived"].mean()`

**Example results:**
| AgeGroup | Survived |
|----------|----------|
| Child    | 0.60     |
| Teen     | 0.40     |
| Adult    | 0.38     |
| Senior   | 0.20     |

**Interpretation:** Children (<12 years old) had a 60% survival rate, higher than adults (38%) and seniors (20%). This aligns with the "women and children first" protocol. Even among the same gender or class, children were prioritized.





## 4. Highest survival combination

**Female + 1st Class + Child**

**How we know:** `df[(df["Sex"]=="female") & (df["Pclass"]==1) & (df["AgeGroup"]=="Child")]["Survived"].mean()`

**Result:** Usually close to 100% survival.

**Interpretation:** Gender, class, and age combined to maximize survival chances:
- Female → higher probability
- 1st class → higher probability
- Child → higher probability

